### PR TITLE
RFC: Tune end-to-end ML workflows in KerasTuner

### PR DESCRIPTION
**Pull request by haifeng-jin**

### CODE REVIEW

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+    outputs = keras.layers.Dense(1, activation='relu')(x)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    model.compile(loss='mse',
+                  metrics=['mae'])
+    return model
+  
+  def fit(self, hp, model, dataset=None, callbacks=None, **kwargs):
+    # Data preprocessing
+    # Get the hyperparameter value used in `build()`.
+    dataset = dataset.map(hp.get("num_features"))
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: This should be an actual data transformation

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+models = tuner.get_best_models(num_models=2)
+
+# Retraining the model with the second best hyperparameters.
+second_best_hp = tuner.get_best_hyperparameters(num_models=2)[1]
+hypermodel = MyHyperModel()
+model = hypermodel.build(second_best_hp)
+hypermodel.fit(model, hp=second_best_hp, callbacks=[], dataset=new_dataset)
+```
+
+There are three details in this workflow that are worth attention.
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: "Please take note of the following four points:"

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+    return model
+  
+  def fit(self, hp, model, x, y, validation_data, callbacks, **kwargs):
+    # Create and adapt the text vectorizer.
+    text_vectorizer = layers.TextVectorization(
+        # The max_tokens is a hyperparameter created in build().
+        max_tokens=hp.get("max_tokens"),
+        output_mode="int",
+        output_sequence_length=10)
+    text_vectorizer.adapt(x)
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: Recomputing the vocab all the time isn't the best. Maybe we should just load a truncated version of the vocab file.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+  
+    # Save the text vectorization model.
+    keras.Sequential([text_vectorizer]).save(
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: You'd just save the vocab file as a text file, not a Sequential.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+        os.path.join(self.trial_dir, "text_vectorization_model"))
+  
+    return model.fit(
+        # Convert x from strings to integer vectors.
+        text_vectorizer(x),
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: Use `map(lambda x, y: (text_vectorizer(x), y), num_parallel_calls=tf.data.AUTOTUNE)`

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+        run_val_step(data)
+      val_loss = val_loss_tracker.result().numpy()
+      min_val_loss = min(min_val_loss, val_loss)
+      print(f"Epoch val_loss: {val_loss}")
+      val_loss_tracker.reset_states()
+  
+    return min_val_loss
+```
+
+You may also subclass `keras.Model` to override `train_step`.
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: Refer to method names with parens, e.g. `train_step()`

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+
+#### Fine tuning with pretrained weights
+
+```py
+class FineTuneModel(keras.Model):
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: Prefer removing this class

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+class MyHyperModel(kt.HyperModel):
+
+  def build(self, hp):
+    head = keras.Sequential([
+      layers.GlobalAveragePooling2D(),
+      layers.Dense(hp.Int("units", 32, 128)),
+      layers.Dense(1),
+    ])
+    return FineTuneModel(
+        base_model=keras.applications.ResNet50(
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: You can just make the ResNet the first layer in the Sequential (or it could be a functional model as well)

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+            weights="imagenet",
+            input_shape=(32, 32, 3),
+            include_top=False,
+        ),
+        head=head)
+  
+  def fit(self, hp, model, dataset, validation_data, callbacks, **kwargs):
+    # Fit the model with the `base_model` freezed.
+    model.base_model.trainable = False
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: Then you just do model.layers[0].trainable = False (or whichever index is the base model)

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+    dataset = dataset.map(hp.get("num_features"))
+    # Model training
+    # Returning the training history
+    # or a similar dictionary if using custom training loop.
+    return model.fit(
+        dataset,
+        epochs=100,
+        validation_data=validation_data,
+        # Tune whether to use shuffle.
+        shuffle=hp.Boolean("shuffle"),
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:302]: If it's a dataset, you can't use this arg. Shuffling happens on the dataset object.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+        # Tune whether to use sample_weights.
+        sample_weight=sample_weight if hp.Boolean("sample_weight") else None,
```
Comment (fchollet) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:304]: Likewise this should be managed by the dataset

**Review by fchollet — COMMENTED** [review]
Looks great!

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 10 above)
Comment (haifeng-jin) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:302]: Changed to not using datasets to show case tuning fit args.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 5 above)
Comment (haifeng-jin) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: So we use `tf.data.Dataset` and `text_vectorizer.adapt(dataset.map(lambda x, y: x))`?

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 4 above)
Comment (haifeng-jin) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: @mattdangerw Is there any code examples for saving, loading, and truncating the vocab instead of the layer?

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 12]
```diff
@@ -0,0 +1,535 @@
...
+data, they need to override the `Tuner.run_trial()` function, which adds more
+to the learning curve.
+
+For example, in natural language processing, tokenization and vectorization may
+affect the later model type. They will need to find a way to pass this
+information from `Tuner.run_trial()` to HyperModel.build.
+
+#### Tune existing Keras code
+
+If the users have their code of model building and training ready written using
```
Comment (mattdangerw) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: code of -> code for

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 5 above)
Comment (mattdangerw) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: Yeah, that looks right.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 4 above)
Comment (mattdangerw) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md]: Hmm, you could save a vocabulary for a TextVectorization to a file like this.

```
vocab = text_vectorizer.get_vocabulary(include_special_tokens=False)
with open('vocab.txt', 'w') as file:
  for term in vocab:
    file.write(term + '\n')
```

But the issue is we don't really support loading vocabulary file only up to a truncated line. I don't believe the underlying op supports this either though would need to check.

Things that would work today.
 1) save file, truncate file by hand after saving (maybe awkward)
 2) truncate vocabulary by creating new TextVectorization layers
 3) truncate only output of layer

```
adapted_tv = layers.TextVectorization()
adapted_tv.adapt(data)
# option 2, truncate by creating new layer
vocab = adapted_tv.get_vocabulary()
truncated_tv = layers.TextVectorization(vocabulary=vocab[:max_tokens])

# option 3, truncate output from layer
def apply_tv(input):
  output = adapted_tv(input)
  # Truncate layer output to smaller vocab size. 1 is the oov token.
  return tf.where(output >= max_tokens, tf.ones_like(output), output)
```

Would any of those make sense? Seems like truncating layer output might be easiest.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+Reduce the learning curve and code hacks for workflows involves hyperparameters
+in data preprocessing and model fitting.
+
+## Motivation
+
+Different users prefer different workflows for their tuning process -- like
+Keras has different getting-started tutorials for engineers and researchers.
+There are users who prefer to learn more about the framework and to implement
+everything by overriding class methods, and users who prefer to write
+everything from scratch to have a shorter learning curve and better
```
Comment (sayakpaul) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:20]: Shouldn't it be a longer learning curve when implementing things from scratch?

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+configurability for the details.  For example, some users would like to
+override `Model.train_step()` to make the code cleaner, others like to write
+the training loop from scratch.
+
+
+Currently, KerasTuner has good support for the users who would like to
+restructure their code by learning the KerasTuner framework, and for users who
+only need to do some light customization of the model building process.
+However, the support for users who need to write their model building and
+training process from scratch is not adequate.
```
Comment (sayakpaul) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:30]: Might be worth mentioning that it's still, however, possible to fuse `Model.train_step()` and Keras Tuner functionalities in a clean manner.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+    model.compile(loss='mse',
+                  metrics=['mae'])
+    # Model training
+    history = model.fit(
+        x,
+        y,
+        epochs=100,
+        validation_data=validation_data,
+        # Tune whether to use shuffle.
+        shuffle=hp.Boolean("shuffle"),
```
Comment (sayakpaul) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:188]: What about `tf.data.Dataset` objects instead of numpy arrays? I suspect as long as `feature_selection()` yields the data in a format accepted by Keras, it should be fine? I am asking this irrespective of the `shuffle` argument. 

If so, I think it's worth specifying that.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md [hunk 16]
```diff
@@ -0,0 +1,555 @@
...
+       ...
+       trial.trial_id,
+   ))
+   # Save model can be handled by the user.
+   # `trial_id` is unique for each trial.
+
+tuner = MyTuner(...)
+tuner.search()
+# Retraining the model
+build_and_evaluate_model(**tuner.get_best_hyperparameters()[0])
```
Comment (sayakpaul) [inline rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:257]: Might be worth mentioning that `build_and_evaluate_model` is a utility and then appending a brief description about its inputs and what it does.

**Review by sayakpaul — CHANGES_REQUESTED** [review]
Looking forward to this so much.

**Comment by bhack** [comment]
I don't see any example here in this refactoring about the model arch itself (Automodel/[Customized Automodel](https://autokeras.com/tutorial/customized/)).

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 13 above)
Comment (haifeng-jin) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:20]: A shorter learning curve refers to the learning curve for KerasTuner.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 14 above)
Comment (haifeng-jin) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:30]: Any suggested workflow for that?

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 15 above)
Comment (haifeng-jin) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:188]: Yes, however, the `shuffle` arg only works for non-tf.data.Dataset type. For `tf.data.Dataset`, we can use `tf.data.Dataset.shuffle()`.

Comment (haifeng-jin) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:188]: Sure. Later in the examples, we have the NLP case using `tf.data`.

**Comment by haifeng-jin** [comment]
> I don't see any example here in this refactoring about the model arch itself (Automodel/[Customized Automodel](https://autokeras.com/tutorial/customized/)).

This is a proposed API change for KerasTuner.
Any change, if needed, related to `AutoModel` in AutoKeras would be in another design doc.

**Comment by bhack** [comment]
> This is a proposed API change for KerasTuner.
> Any change, if needed, related to `AutoModel` in AutoKeras would be in another design doc.

Ok, please if you can notify us in the forum when we will have also this design doc.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 13 above)
Comment (sayakpaul) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:20]: Still sounds a bit confusing TBH.

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 14 above)
Comment (sayakpaul) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:30]: Yeah so, as long as one is able to define how the loss is calculated and the parameters are updated I think the standard KerasTuner workflows would work. Consider this example: https://keras.io/examples/nlp/nl_image_search/. We could likely do a search with the regular KerasTuner workflows to optimize the hyperparameters of `DualEncoder`. Please correct me if I am wrong?

File: rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md (hunk 15 above)
Comment (sayakpaul) [reply rfcs/20210920-tune-end-to-end-ml-workflows-in-keras-tuner.md:188]: Of course. Might be worth specifying this if not already done.
